### PR TITLE
State: Use go-datastore to implement the state interface

### PR DESCRIPTION
**Pull request by hsanjuan**
Since the beginning, we have used a Go map to store the shared state (pinset)
in memory. The mapstate knew how to serialize itself so that libp2p-raft would
know how to write to disk when it:

* Saved snapshots of the state on shutdown
* Sent the state to a newcomer peer

hashicorp.Raft assumes an in-memory state which is snapshotted from time to
time and read from disk on boot.

This commit adds a `dsstate` implementation of the state interface using
`go-datastore`. This allows to effortlessly switch to a disk-backed state in
the future (as we will need), and also have at our disposal the different
implementations and utilities of Datastore for fine-tuning (caching, batching
etc.).

`mapstate` has been reworked to use dsstate. Ideally, we would not even need
`mapstate`, as it would suffice to initialize `dsstate` with a
`MapDatastore`. BUT, we still need it separate to be able to auto-migrate to
the new format.

This will be the last migration with the current system. Once this has been
released and users have been able to upgrade we will just remove `mapstate` as
it is now.

License: MIT
Signed-off-by: Hector Sanjuan <code@hector.link>

Depends on https://github.com/libp2p/go-libp2p-raft/pull/50

### CODE REVIEW

**Comment by hsanjuan** [comment]
Also, kind of depends on discussion #654 . If we slightly changed how things are  serialized soon, I'd rather not have to migrate right after this.

**Comment by lanzafame** [comment]
LGTM. Only thing would be a better explanation of what the migration process is and how the migration process will be different in the future, i.e. post v6.

**Comment by lanzafame** [comment]
@hsanjuan It seems that in some cases, this PR has slowed down cluster as there have now been 2 test timeout failures. I am happy to bump the timeout up again and then chip away at getting it down again.

**Comment by hsanjuan** [comment]
> Also, kind of depends on discussion #654 . If we slightly changed how things are serialized soon, I'd rather not have to migrate right after this.

I don't fully know. But since checking state version will mean opening the datastore and reading one key, we know that will be different. The migration will be looping the keys and overwriting them with their updated form.

The key idea is that Marshal/Unmarshal will either not be used, or can happen normally regardless of the state version (unlike now).

**Comment by hsanjuan** [comment]
> @hsanjuan It seems that in some cases, this PR has slowed down cluster as there have now been 2 test timeout failures. I am happy to bump the timeout up again and then chip away at getting it down again.

I only see jenkins, which I retriggered now. It make take a while before I merge this anyway.

Does this approval include changes in go-libp2p-raft ?

**Comment by lanzafame** [comment]
@hsanjuan I just added some comments on the go-libp2p-raft PR.

**Comment by hsanjuan** [comment]
@lanzafame do me a favour and have a look to the last commits (added tracing) and lmk if we can leave this ready for merge, as I want it to go before #670 .

File: cluster.go [hunk 1]
```diff
@@ -739,20 +739,7 @@ func (c *Cluster) Join(ctx context.Context, addr ma.Multiaddr) error {
...
-		if err != nil {
-			logger.Error(err)
-		}
-		ch <- time.Now() // boostrap
-		defer close(ch)
-		select {
-		case <-dhtBstCtx.Done(): // shut down the process
-			proc.Close()
-		}
+		c.dht.BootstrapOnce(c.ctx, dht.DefaultBootstrapConfig)
```
Comment (lanzafame) [inline cluster.go]:
```suggestion
		c.dht.BootstrapOnce(ctx, dht.DefaultBootstrapConfig)
```
`ctx` has already been derived from the `c.ctx` at the top of the function.

File: api/types.go [hunk 2]
```diff
@@ -704,10 +707,11 @@ func (pT PinType) String() string {
 
 // PinOptions wraps user-defined options for Pins
 type PinOptions struct {
-	ReplicationFactorMin int    `json:"replication_factor_min"`
-	ReplicationFactorMax int    `json:"replication_factor_max"`
-	Name                 string `json:"name"`
-	ShardSize            uint64 `json:"shard_size"`
+	ReplicationFactorMin int               `json:"replication_factor_min" codec:"rn,omitempty"`
```
Comment (lanzafame) [inline api/types.go]: @hsanjuan Is there a compile-time or test-time check that we can add so that we don't conflict with the codec variable?

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 1491a01ec8028fd4e593bcff6a33cf66cce6d060
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by:**

Code contribution guidelines for IPFS Cluster are available at https://cluster.ipfs.io/developer/contribute/#code-contribution-guidelines

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 1491a01ec8028fd4e593bcff6a33cf66cce6d060
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by:**

Code contribution guidelines for IPFS Cluster are available at https://cluster.ipfs.io/developer/contribute/#code-contribution-guidelines

---
This message was auto-generated by https://gitcop.com

File: api/pb/types.proto [hunk 3]
```diff
@@ -0,0 +1,23 @@
+syntax = "proto3";
+package api.pb;
+
+message Pin {
+  bytes Cid = 1;
+  enum PinType {
```
[the remark this answers is not in the document]
Comment (hsanjuan) [reply api/pb/types.proto]: I can't have an enum called `Type` and a field called `Type`

File: api/types.go (hunk 2 above)
Comment (hsanjuan) [reply api/types.go]: As discussed, tests break in most cases when screwing up the codec field names.

File: api/pb/types.proto (hunk 3 above)
[the remark this answers is not in the document]
Comment (hsanjuan) [reply api/pb/types.proto]: let's leave this as is, I don't mind so much and we can change it later.

File: api/types.go (hunk 2 above)
Comment (hsanjuan) [reply api/types.go]: Additionally I have added a test that verifies that there are no dup fields.

**Comment by hsanjuan** [comment]
@lanzafame I squashed, rebased and cleaned up. Then I tested with larger states and had to fix some issues (latest commit). Can you check again? I think we can merge then.

**Comment by lanzafame** [comment]
@hsanjuan LGTM

**Pull request merged**